validatorobserver/validator_observer.go

```diff
@@ -0,0 +1,197 @@
+	if err != nil {
+		registeredChains = []contractbinding.IBridgeStructsChain{}
+	}
+
+	for _, chain := range registeredChains {
+		validatorsData, err := bridgeSmartContract.GetValidatorsChainData(context.Background(),
+			common.ToStrChainID(chain.Id))
+		if err != nil {
+			continue
+		}
+
+		validatorKeys := []eth.ValidatorChainData{}
+		for _, data := range validatorsData {
+			validatorKeys = append(validatorKeys, eth.ValidatorChainData{
+				Key: data.Key,
+			})
+		}
+
+		slotNumber := uint64(0)
+
+		lastObservedBlock, err := bridgeSmartContract.GetLastObservedBlock(ctx, common.ToStrChainID(chain.Id))
+		if err == nil {
+			slotNumber = lastObservedBlock.BlockSlot.Uint64()
+		}
+
+		validators.Data[common.ToStrChainID(chain.Id)] = ValidatorsChainData{
+			Keys:       validatorKeys,
+			SlotNumber: slotNumber,
+		}
+	}
+
+	return &ValidatorSetObserver{
+		validatorSetStream:  make(chan *Validators),
+		validators:          validators,
+		validatorSetPending: validatorSetPending,
+		bridgeSmartContract: bridgeSmartContract,
+		logger:              logger.Named("validator_set_observer"),
+	}, nil
+}
+
+func (vs *ValidatorSetObserver) Start(ctx context.Context) {
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				close(vs.validatorSetStream)
+
+				return
+			case <-time.After(timeout):
+				if err := vs.execute(); err != nil {
+					vs.logger.Error("error while executing", "err", err)
+				}
+			}
+		}
+	}()
+}
+
+func (vs *ValidatorSetObserver) execute() error {
+	lock.Lock()
+	oldState := vs.validatorSetPending
+	validators := vs.validators
+	lock.Unlock()
+
+	isPending, err := vs.bridgeSmartContract.IsNewValidatorSetPending()
+	if err != nil {
+		return err
+	}
+
+	if isPending == oldState {
+		return nil
+	}
+
+	if isPending {
+		addedValidators, removedValidators, err := vs.bridgeSmartContract.GetVerificationKeys()
+		if err != nil {
+			return err
+		}
+
+		vs.removeValidators(validators, removedValidators)
+		vs.addValidators(validators, addedValidators)
+	}
+
+	lock.Lock()
+	defer lock.Unlock()
+
+	vs.validatorSetPending = isPending
+	if isPending {
+		vs.validators = validators
+		vs.validatorSetStream <- &validators
+	} else {
+		vs.validatorSetStream <- nil
+	}
+
+	return nil
+}
+
+func (vs *ValidatorSetObserver) IsValidatorSetPending() bool {
+	lock.RLock()
+	defer lock.RUnlock()
+
+	return vs.validatorSetPending
+}
+
+func (vs *ValidatorSetObserver) GetVerificationKeys(chainID string) []eth.ValidatorChainData {
+	lock.RLock()
+	defer lock.RUnlock()
+
+	return vs.validators.Data[chainID].Keys
+}
+
+func (vs *ValidatorSetObserver) GetValidatorSetReader() <-chan *Validators {
+	return vs.validatorSetStream
+}
+
+func (vs *ValidatorSetObserver) removeValidators(validators Validators,
+	removedValidators []ethcommon.Address) {
+	for _, validator := range removedValidators {
+		validatorIdx, err := vs.bridgeSmartContract.GetAddressValidatorIndex(validator)
```

Comment:
do not use smart contract for this. Validators are already sorted by their indexes.

We only need one loop and previously generated map here